### PR TITLE
feat: Select market to support track relinking

### PR DESCRIPTION
**Pull request by grufkork**
Added configuration option to select market, which allows for following Relinked tracks. Solves #75. Not sure if this is the way you'd like to implement it, but it's a fix before the rewrite drops!

### CODE REVIEW

**Comment by oSumAtrIX** [comment]
Is it not possible to automatically iterate over (available) markets? What is the necessity to expose this choice to the user?

**Comment by grufkork** [comment]
I'm not quite sure I understand what you mean. I suppose you could just try every market until you get a hit, but Spotify's API only gives you a redirect if you give it an explicit market to look in. You can't do a "give me the relinks of this song", but have to actually try to get the song. Choosing the user's local market should lead to the most reliable/expected behaviour. I tried using the market auto detection, but the program just exited, and I didn't investigate further. 

Exposing the choice is a bit ugly, but it might be a better idea than trying to auto-detect and then possibly having different song availability than what the user normally has in the Spotify app.

Thanks for the great program by the way!

File: src/settings.rs [hunk 1]
```diff
@@ -50,6 +52,7 @@ impl Settings {
 			client_secret: client_secret.to_string(),
 			refresh_ui_seconds: 1,
 			downloader: DownloaderConfig::new(),
+			market: Some(CountryCode::USA),
```
Comment (oSumAtrIX) [inline src/settings.rs:55]: Is USA the default market in librespot?

**Comment by oSumAtrIX** [comment]
By automatically choosing the market I meant doing what is described here:

image

**Comment by grufkork** [comment]
I think it does the same thing, except the Spotify app knows the users market from their account. There's a `Market::FromToken` which should auto-select, but it didn't work for me when I tried it. It might be that it requires further permissions. As for CountryCode default, I don't think there's a default. Not sure what market Spotify uses when it isn't supplied. 

I however can't test more right now as suddenly *nothing* is downloading. Not sure if I've gotten my API access restricted, it happened suddenly mid-way through a download.

**Comment by oSumAtrIX** [comment]
"None" was the default:

image

For that reason, keep the default none as well. If necessary, the user can then choose a country code, but by default, it should behave like before this PR.

I am still conflicted about having the user choose this. Spotify does it automatically, and we should, too. The user shouldn't be involved in such a technicality. As I understand it, the market has to be chosen for each download. One download might need one market and another, a different one; this implementation does not allow that, furthermore speaking for choosing the market dynamically.

**Comment by grufkork** [comment]
Oh right, I see. I'll fix that - just removing it from the default config should work. 

I do however think it would be  good idea to add an option/flag to do that kind of market scanning. As far as I know, you can only check if a song is relinked in a given market, not get a list of which markets a song is relinked in. You'd thus have to try each and every market to see if it exists. I think `Market::FromToken` should provide the same behaviour as regular Spotify, I'll look into that.

**Comment by grufkork** [comment]
Default fixed. However got this error using FromToken
image

**Comment by grufkork** [comment]
Someone [here](https://community.spotify.com/t5/Spotify-for-Developers/Web-API-artist-top-tracks-endpoint-is-requiring-quot-market-quot/td-p/5570312) says you need to specify market when using client secret authorisation. Also not sure why the check failed, I don't think it's something I did?

**Comment by oSumAtrIX** [comment]
So from the API docs, the idea behind markets is the following:

image

Currently, DownOnSpot will automatically select a market to download the track from which is why I am slightly confused why this PR is necessary:

https://github.com/oSumAtrIX/DownOnSpot/blob/main/src/downloader.rs#L498-L500

**Comment by grufkork** [comment]
I haven't figured out what the `find_alternative` method does, but I think it only tries a different market on the same song, assuming it exists. If you look up the metadata for https://open.spotify.com/track/7lds8jOEJM9d72i0A0YEfX [eg. on this](https://metagrab.barkloaf.com/), you'll see it has 0 available markets as it has been replaced by a completely new song. Trying to download it with DownOnSpot fails, unless you give it a market, at which point Spotify will instead automatically give you the new track. 

In other words, I think the current implementation works for tracks that are still available, just not in your current market. It does however not work for tracks which have been Relinked, where the old version is completely removed.

**Comment by raffomania** [comment]
I think `find_alternatives` has a typo in [this line](https://github.com/oSumAtrIX/DownOnSpot/blob/92cf74bdeda7632b367fb8561e5b01d2ffce1ebe/src/downloader.rs#L477). It should be `if track.available`, without the `!`.

**Comment by oSumAtrIX** [comment]
@raffomania That's a good find. I'll push for a fix right away. Thanks!

>  It might still be necessary for some scenarios, depending on where librespot's "alternatives" come from.

I believe the alternatives are what Spotify returns in the API. 

> the "preferred" way to automatically handle this is to read the user's profile to find out the correct market parameter to set.

I have yet to understand why that is the case. What's a scenario where one would need to specify a market?

**Comment by grufkork** [comment]
I'm not sure, but I think the relinked track has to exist in the market that was requested for the old song.

**Comment by oSumAtrIX** [comment]
In this case, this sounds like a case-by-case scenario where you have to figure out the market for each song individually, otherwise, if you set a global market for all songs, some may be available in another.

**Comment by grufkork** [comment]
Almost all songs are globally available, but the bottom line is that users expect to be able to download the songs they can play on regular Spotify. Setting the market to your country matches the expected behaviour (it's what Spotify does), and anyone who is fudging the market on their account likely knows enough to manage the settings as they'd like.

**Comment by oSumAtrIX** [comment]
I am primarily using DownOnSpot to download songs from YouTube. I do not use Spotify for any other reason, so making this assumption that downloading from the market the account is set to be enough would break the functionality in the worst case for me. 

I have checked the track you sent earlier, and the available markets array is empty regardless of whether you give it a market or not: https://open.spotify.com/track/7lds8jOEJM9d72i0A0YEfX

Unless a market was specified, though, the availability of the track was false. Do you have an example of a song that is unavailable in a specific market and has an empty available market array?

**Comment by grufkork** [comment]
Right, fair point. 

So you have tracks like this one, which have been relinked: https://open.spotify.com/track/7lds8jOEJM9d72i0A0YEfX. When you send a request via the API, you either get back the track saying it doesn't exist, or if you filled out the market parameter, you get the new track directly and a notice that you have been redirected.

Then you have tracks like this: https://open.spotify.com/track/0UyQ9TuZ1lG6eJi4eRKDfO which have been completely removed. When you make a request, you get back the original track regardless of market parameter.

I don't have much more time to put into this, but this PR made it so that I could download 99.9% of ~1000 tracks, whereas before maybe 1/20 would fail (due to the playlists being old and having outdated links to songs). I'd say add the feature as it is, and add a notice that the user might want to experiment with different markets. At any rate, it does not reduce the number of tracks that will be downloadable.

**Comment by ValentinFunk** [comment]
Thanks @grufkork!

File: src/spotify.rs [hunk 2]
```diff
@@ -43,7 +45,11 @@ impl Spotify {
 		};
 		let spotify = Client::new(credentials);
 
-		Ok(Spotify { session, spotify })
+		Ok(Spotify {
+			session,
+			spotify,
+			market,
```
Comment (oSumAtrIX) [inline src/spotify.rs:51]: Is it possible here to use the Spotify API via the aspotify client to get the market of the account automatically?

File: src/main.rs [hunk 3]
```diff
@@ -80,6 +81,11 @@ async fn start() {
 		&settings.password,
 		&settings.client_id,
 		&settings.client_secret,
+		if let Some(countrycode) = settings.market {
+			Some(Market::Country(countrycode))
+		} else {
+			None
+		},
```
Comment (oSumAtrIX) [inline src/main.rs:88]: This logic could be moved into `Spotify::new` so that in this line we just pass the setting just like the other settings.

File: src/spotify.rs (hunk 2 above)
Comment (grufkork) [reply src/spotify.rs:51]: Was a while since I looked at this, but from what I remember, the kind of Spotify development app that is used is largely separate from the user's account. Thus it isn't really associated with a region/market. There's probably lookups that could be made separately, but nothing right out of the box. There's a `Market::from_token` but it failed last time I tried: https://github.com/oSumAtrIX/DownOnSpot/pull/76#issuecomment-1958116115

File: src/settings.rs [hunk 4]
```diff
@@ -22,6 +23,7 @@ pub struct Settings {
 	pub client_secret: String,
 	pub refresh_ui_seconds: u64,
 	pub downloader: DownloaderConfig,
+	pub market: Option<CountryCode>,
```
Comment (oSumAtrIX) [inline src/settings.rs:26]: Maybe call this `market_country_code` so it's clear in the JSON that a country code is expected. The user will infer that they should use their own countries code.

File: src/main.rs (hunk 3 above)
Comment (grufkork) [reply src/main.rs]: Moved!

File: src/settings.rs (hunk 4 above)
Comment (grufkork) [reply src/settings.rs]: Agree, renamed.

**Review by oSumAtrIX — APPROVED** [review]

**Comment by oSumAtrIX** [comment]
Thanks!

**Comment by grufkork** [comment]
Oh, one thing to consider might be what happens to current configs. Without the option there will simply happen nothing, but maybe it should be automatically added to existing configs. Possibly some docs need updating too?

**Comment by oSumAtrIX** [comment]
I am not sure if there's a universal way to keep the config up to date. Otherwise just supporting this specific case when in the future there might be similar ones would be a problem.
